File: C#/prime.cs
using System;
using System.Linq;
using System.Collections.Generic;

class Prime {
	
	// returns true/false for if n is prime
	public static bool prime(int n) {
		if (n <= 1) {
			return false;
		}
		if (n == 2) {
			return true;
		} else {
			int k = 2;
			return primeRec(n, k);
		}
	}

	// recursive function for prime
	public static bool primeRec(int n, int k) {
		// if divisible then not prime
		if ((n % k) == 0) {
			return false;
		}
		// a limit to stop (correct limit would be sqrt(n))
		if ((n / 2) < k) {
			return true;
		// else check next divisor
		} else {
			return primeRec(n, k + 1);
		}
	}

	// implementation of sieve of erotosthanes
	// returns an array of truth values marking whether an index is prime
	public static bool[] sieve(int n) {
		// correct limit of sqrt(n)
		double lim = Math.Sqrt(Convert.ToDouble(n));
		int intLim = Convert.ToInt32(lim);
		int i = 2;
		bool[] bools = Enumerable.Repeat(true, n).ToArray();
		bools[0] = false;
		bools[1] = false;
		while (i < intLim) {
			if (bools[i] == true) {
				int j = i * i;
				while (j < n) {
					bools[j] = false;
					j = j + i;
				}
			}
			i = i + 1;
		}
		return bools;
	}
	
	// yields the next prime from k to n
	// IEnumberable allows for yielding
	public static IEnumerable<int> sievePrimes(int n, int k) {
		bool[] bools = sieve(n);
		while (k < n) {
			if (bools[k]) {
				yield return k;
			}
			k = k + 1;
		}
	}

	// yields the next prime partition as a LinkedList
	// IEnumberable allows for yielding
	public static IEnumerable<LinkedList<int>> primePartition(int n, int k) {
		// bad input then return nothing
		if (n <= 1 || n < k) {
			yield break;
		}
		// if prime yield n
		if (prime(n)) {
			LinkedList<int> lst = new LinkedList<int>();
			lst.AddFirst(n);
			yield return lst;
		}
		// iterate through the possible primes
		IEnumerable<int> primes = sievePrimes(n, k);
		foreach (var i in primes) {
			// iterate through the yielded partitions of subproblems
			IEnumerable<LinkedList<int>> partitions = primePartition(n - i, i + 1);
			foreach (var j in partitions) {
				// add the prime to the sub prime partition
				j.AddFirst(i);
				yield return j;
			}
		}
	}
	
	// simple implementation of a LinkedList printing function
	public static void printLinkedList(LinkedList<int> lst) {
		if (lst == null) {
			return;
		}
		if (lst.First == null) {
			return;
		}
		LinkedListNode<int> tmp = lst.First;
		while (tmp != null) {
			Console.Write(tmp.Value);
			Console.Write(" ");
			tmp = tmp.Next;
		}
		Console.Write("\n");
	}

	// REPL
	static void Main(string[] args) {
		int n = 1;
		while (true) {
			// catches improper input (at least strings)
			try {
				Console.WriteLine("Please enter a positive integer N for which you would like the distinct prime partitions (0 to quit): ");
				n = Convert.ToInt32(Console.ReadLine());
			} catch (Exception e) {
				Console.WriteLine(e.Message);
				Console.WriteLine("You most likely didn't type a number, I asked for a number, please input a number... ");
				continue;
			}
			// terminates loop on 0
			if (n == 0) {
				break;
			// retries on negative
			} else if (n < 0) {
				Console.WriteLine("I thought I made it pretty clear that you need to enter a positive integer... ");
			// else call the function and iterate and print through yields
			} else {
				IEnumerable<LinkedList<int>> partitions = primePartition(n, 1);
				foreach (var i in partitions) {
					printLinkedList(i);
				}
			}
		}
	}

}

File: Python/prime.py
# function returning the bool for whether n is prime, recursive
def prime(n):
    if n <= 1:
        return False
    if n == 2:
        return True
    else:
        k = 2
        return primeRec(n, k)

# recursive function for the prime(n) function
def primeRec(n, k):
    # if divisible then return not prime
    if (n % k) == 0:
        return False
    # sqrt(n) is the upper bound of numbers that can divide
    if (n**(1/2)) < k:
        return True
    # else check the next possible divisor
    else:
        return primeRec(n, k + 1)

# implementation of sieve of erotosthanes
# returns an array of bools marking whether an index is prime
def sieve(n):
    # largest number that can divide a number <= n
    lim = (n**(1/2))
    i = 2
    bools = [True] * (n + 1)
    bools[0] = False
    bools[1] = False
    while (i < lim):
        if (bools[i] == True):
            j = i**2
            while (j <= n):
                bools[j] = False
                j = j + i
        i = i + 1
    return bools

# yields the next prime from k to n
def sievePrimes(n, k):
    bools = sieve(n)
    while (k <= n):
        if bools[k]:
            yield k
        k = k + 1

# recursively yields the prime partitions as arrays
def primePartition(n, k):
    # not a good input
    if (n <= 1 or n < k):
        return
    # if n is prime then yield it
    if (prime(n)):
        yield [n]
    # iterate through possible primes
    for i in sievePrimes(n, k):
        # iterate through possible paritions
        for j in primePartition(n - i, i + 1):
            # add the prime to the partition it was subtracted from
            lst = [i]
            lst.extend(j)
            yield lst

# starts the call to primePartition(n, 1) and prints
def startPrimePartition(n):
    for i in primePartition(n, 1):
        print(i)

# REPL
def main():
    while True:
        print("Please enter a positive integer N for which you would like the distinct prime partitions (0 to quit): ")
        # catches improper inputs (at least strings)
        try:
            n = int(input())
        except ValueError:
            print("Try again, an integer is a number if you didn't know...")
            continue;
        # stops loop
        if (n == 0):
            break;
        # negative integer
        elif (n < 0):
            print("Try again, a POSITIVE number this time... ")
        # result
        else:
            startPrimePartition(n)

if __name__ == "__main__":
    main()
